habitat-lab/habitat/tasks/rearrange/multi_task/pddl_defined_predicates.py

```diff
@@ -0,0 +1,384 @@
+from typing import Optional, cast
+
+import magnum as mn
+import numpy as np
+
+import habitat_sim
+from habitat.sims.habitat_simulator.sim_utilities import get_ao_global_bb
+from habitat.tasks.rearrange.marker_info import MarkerInfo
+from habitat.tasks.rearrange.multi_task.rearrange_pddl import (
+    PddlEntity,
+    PddlSimInfo,
+    SimulatorObjectType,
+)
+from habitat.tasks.rearrange.utils import (
+    place_agent_at_dist_from_pos,
+    rearrange_logger,
+)
+
+# TODO: Deprecate these and instead represent them as articulated object entity type.
+CAB_TYPE = "cab_type"
+FRIDGE_TYPE = "fridge_type"
+
+
+def is_robot_hold_match(
+    robot: PddlEntity,
+    sim_info: PddlSimInfo,
+    hold_state: bool,
+    obj: Optional[PddlEntity] = None,
+) -> bool:
+    """
+    Check if the robot is holding the desired object in the desired hold state.
+    :param hold_state: True if the robot should be holding the object.
+    """
+
+    robot_id = cast(
+        int,
+        sim_info.search_for_entity(robot),
+    )
+    grasp_mgr = sim_info.sim.get_agent_data(robot_id).grasp_mgr
+
+    if hold_state:
+        if obj is not None:
+            # Robot must hold specific object.
+            obj_idx = cast(int, sim_info.search_for_entity(obj))
+            abs_obj_id = sim_info.sim.scene_obj_ids[obj_idx]
+            return grasp_mgr.snap_idx == abs_obj_id
+        else:
+            # Robot can hold any object.
+            return grasp_mgr.snap_idx != None
+    else:
+        # Robot must hold no object.
+        return grasp_mgr.snap_idx == None
+
+
+def set_robot_holding(
+    robot: PddlEntity,
+    sim_info: PddlSimInfo,
+    hold_state: bool,
+    obj: Optional[PddlEntity] = None,
+) -> None:
+    robot_id = cast(
+        int,
+        sim_info.search_for_entity(robot),
+    )
+    sim = sim_info.sim
+    agent_data = sim.get_agent_data(robot_id)
+    # Set the snapped object information
+    if not hold_state and agent_data.grasp_mgr.is_grasped:
+        agent_data.grasp_mgr.desnap(True)
+    elif hold_state:
+        # Swap objects to the desired object.
+        obj_idx = cast(int, sim_info.search_for_entity(obj))
```

Comment:
what if obj is None?

Reply:
I added an error message if this is the case.